### PR TITLE
Implement endpoint bindings for applications (#400)

### PR DESCRIPTION
**Pull request by gboutry**
## Description

The goal is to enable binding endpoints to specific spaces. This features will used in sunbeam when deploying with MAAS.

Fixes: #400 

## Type of change

*\<What type of a change is this? Please keep one or more relevant lines from below and delete the others.\>*

- Change existing resource
- Logic changes in resources (the API interaction with Juju has been changed)
- Change in tests (one or several tests have been changed)
- Requires a documentation update
- Maintenance work (repository related, like Github actions, or revving the Go version, etc.)

## Environment

- Juju controller version: 3.4

- Terraform version: 1.6

## QA steps

Manual QA steps should be done to test this PR.

- You need to have the spaces `management` and `public` associated to the model
```bash
lxc network create management-br ipv4.address=10.150.40.1/24 ipv4.nat=true ipv6.address=none ipv6.nat=false
lxc network create public-br ipv4.address=10.170.80.1/24 ipv4.nat=true ipv6.address=none ipv6.nat=false
juju add-model my-model
juju add-space management 10.150.40.0/24
juju add-space public 10.170.80.0/24
```

```tf
provider juju {}

data "juju_model" "resource" {
  name = "my-model"
}

resource "juju_application" "zk" {
  name  = "zookeeper"
  model = data.juju_model.resource.name

  constraints = "arch=amd64 spaces=management,public"
  charm {
    name    = "zookeeper"
    channel = "3/stable"
    base    = "ubuntu@22.04"
  }

  endpoint_bindings = [
    {
      space = "public"
    },
    {
      endpoint = "zookeeper"
      space    = "management"
    }
  ]
}
```

## Additional notes

*\<Please add relevant notes & information, links to mattermost chats, other related issues/PRs, anything to help understand and QA the PR.\>*
Some tests were added, with a note about changing space setup when #336 is implemented.
Current integration tests have been updated to add 2 lxd bridges as support for the tests.
Tests need `TEST_PUBLIC_BR` and `TEST_MANAGEMENT_BR` containing the cidrs to be run, otherwise, they are skipped.

On behavior:
Specifying a space without an endpoint will make the space the default space (in term of bindings, not touching model configs)

Constraints are in QA/tests to make sure LXD brings the instance with right networking configuration.

### CODE REVIEW

**Comment by gboutry** [comment]
Force push is adding `sudo` in front of the 2 lxd commands creating the networks

File: internal/juju/applications.go [hunk 1]
```diff
@@ -173,21 +175,42 @@ func (input CreateApplicationInput) validateAndTransform() (parsed transformedCr
...
 	parsed.placement = placements
 
+	endpointBindings := map[string]string{}
+	if len(input.EndpointBindings) > 0 {
+		spaceAPIClient := apispaces.NewAPI(conn)
+		knownSpaces, err := spaceAPIClient.ListSpaces()
+		if err != nil {
+			return parsed, err
+		}
+		kwownSpaceNames := set.NewStrings()
```
Comment (hmlanigan) [inline internal/juju/applications.go:185]: s/kwownSpaceNames/knownSpaceNames/

File: internal/juju/applications.go [hunk 2]
```diff
@@ -173,21 +175,42 @@ func (input CreateApplicationInput) validateAndTransform() (parsed transformedCr
 	}
 	parsed.placement = placements
 
+	endpointBindings := map[string]string{}
```
Comment (hmlanigan) [inline internal/juju/applications.go:178]: This functionality is not required when we use the DeployFromRepository API call, the endpoints will be validated by the controller. We're trying to have less business logic in the clients. 

Please move lines 178-195 to legacyDeploy(). Then `parsed.endpointBindings = input.endpointBindings` can be places at the beginning of the method.

Any juju controller version 3.3+ will use DeployFromRepository

**Review by hmlanigan — COMMENTED** [review]
Thank you for the PR, it'll be a nice addition to the provider functionality. 

Here are my initial comments. I haven't had the chance to run initial QA yet. Over all looks good.

I've updated the GitHub action tests, so that the 2.9 tests will work for lxd now. Please rebase as described here: https://github.com/juju/terraform-provider-juju?tab=readme-ov-file#staying-in-sync

File: internal/juju/applications.go (hunk 2 above)
Comment (gboutry) [reply internal/juju/applications.go:178]: The reason I performed this check before hand is because I get a `panic: runtime error: invalid memory address or nil pointer dereference` when giving a space that does not exist.

I will open a bug on Juju

Comment (gboutry) [reply internal/juju/applications.go:178]: https://bugs.launchpad.net/juju/+bug/2055868

**Comment by gboutry** [comment]
Force push is rebasing on main + adding a new commit on top of the previous ones

File: internal/provider/resource_application_test.go [hunk 3]
```diff
@@ -268,6 +277,92 @@ func TestAcc_ResourceApplication_UpgradeProvider(t *testing.T) {
...
+				Config:    testAccResourceApplicationEndpointBindings(modelName, appName, constraints, map[string]string{"": managementSpace, "ubuntu": publicSpace}),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("juju_application.this", "model", modelName),
+					resource.TestCheckResourceAttr("juju_application.this", "endpoint_bindings.#", "2"),
+					resource.TestCheckTypeSetElemNestedAttrs("juju_application.this", "endpoint_bindings.*", map[string]string{"endpoint": "", "space": managementSpace}),
+					resource.TestCheckTypeSetElemNestedAttrs("juju_application.this", "endpoint_bindings.*", map[string]string{"endpoint": "ubuntu", "space": publicSpace}),
+					testCheckEndpointsAreSetToCorrectSpace(modelName, appName, managementSpace, map[string]string{"": managementSpace, "ubuntu": publicSpace}),
+				),
+			},
+		},
```
[the remark this answers is not in the document]
Comment (gboutry) [reply internal/provider/resource_application_test.go:309]: Considering imports, I'm in a bind (pun intended).

Basically, in `computeEndpointBindingsWithoutDefaultValues`, I compare values from the state and the application's actual values to output only the necessary endpoints bindings to save in the state.

Necessary endpoints = either defined in previous state or different from default space
This works fine when updating, but not when importing.

This is because when importing the state is null.

here's an example of plan where it would not work:

consider the application top-ubuntu with endpoints bindings set as `"": "public"`

And this plan:
```tf
resource "juju_application" "ubuntu" {
  name  = "top-ubuntu"
  model = data.juju_model.resource.name

  constraints = "arch=amd64 spaces=management,public"
  charm {
    name    = "jameinel-ubuntu-lite"
    revision = 10
  }

  endpoint_bindings = [
    {
      space = "public"
    }
  ]
}

```

The space will not be stored in state since it's the same as defined in the application.
And I don't really have a way to differentiate what's default and what's provided by the user, because I don't have access to the plan on a read request.



TL;DR:
Can't test for import with endpoint bindings for the same reasons it's not tested for config. --> can't differentiate on import what's different between resource's defaults and user specified values that are same as defaults.
Same behavior as config

[the remark this answers is not in the document]
Comment (hmlanigan) [reply internal/provider/resource_application_test.go:309]: We need to think of this for real life. I'm sure someone will try it. 

We do handle it for config, as there is code to differentiate between default and changed application config value. It might not be tested.

File: internal/juju/applications.go (hunk 2 above)
Comment (hmlanigan) [reply internal/juju/applications.go:178]: Please add a comment that we can remove the checks once the bug is fixed.

File: internal/provider/resource_application_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gboutry) [reply internal/provider/resource_application_test.go:314]: Ok, I used the model's config option `default-space` to be able to compute what's a default value or not. I was reticent to use it, but I don't think there's another way to know if a space is default or not.

**Comment by gboutry** [comment]
While testing for import state, I've had multiple times the failure:
```
=== RUN   TestAcc_ResourceApplication_EndpointBindings
    resource_application_test.go:293: Step 2/2 error running import: ImportStateVerify attributes not equivalent. Difference is shown below. The - symbol indicates attributes missing after import.
        
          map[string]string{
        - 	"placement": "",
        + 	"placement": "0",
          }
--- FAIL: TestAcc_ResourceApplication_EndpointBindings (32.63s)
```

**Comment by hmlanigan** [comment]
@gboutry the placement issue you've seen in test is a known issue. there is also one with ordering of the placement values I might have a fix for.

**Comment by gboutry** [comment]
Thanks for your review @hmlanigan, I fixed this issue.

I ran manual QA steps + whole acceptance test suite on this new commit.

**Review by hmlanigan — APPROVED** [review]
LGTM. I'm sure many people will be happy with this update. Thank you!
